Default scheduler crash if scheduler extender filter returns a not found node

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:
Default scheduler crash if scheduler extender filter returns a not found node

**What you expected to happen**:
Default scheduler just should consider the not found node as not an candidate, instead of crash itself (panic).

**How to reproduce it (as minimally and precisely as possible)**:
1. The cluster only have 1 node: 0.0.0.1
2. One scheduler extender (with nodeCacheCapable enabled) filter accidentally returns a candidate node: 0.0.0.2
3. Default scheduler recieved the candidate nodes: 0.0.0.2 from scheduler extender, but got below error:
```
I0702 03:17:12.628810       1 scheduler.go:453] Attempting to schedule pod: default/scheduler-test-77
I0702 03:17:13.253047       1 trace.go:81] Trace[1695146606]: "Scheduling default/scheduler-test-77" (started: 2019-07-02 03:17:12.6288148 +0000 UTC m=+18.926347501) (total time: 624.1957ms):
Trace[1695146606]: [47.6µs] [47.6µs] Computing predicates
Trace[1695146606]: [624.1802ms] [624.1326ms] Prioritizing
Trace[1695146606]: [624.1957ms] [15.5µs] END
E0702 03:17:13.253123       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:522
/usr/local/go/src/runtime/panic.go:82
/usr/local/go/src/runtime/signal_unix.go:390
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:208
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/scheduler.go:278
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/scheduler.go:457
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x3ced5f9]
```

The generic_scheduler.go:208 is SuggestedHost:  filteredNodes[0].Name:
```
	// When only one node after predicate, just use it.
	if len(filteredNodes) == 1 {
		metrics.SchedulingAlgorithmPriorityEvaluationDuration.Observe(metrics.SinceInSeconds(startPriorityEvalTime))
		metrics.DeprecatedSchedulingAlgorithmPriorityEvaluationDuration.Observe(metrics.SinceInMicroseconds(startPriorityEvalTime))
		return ScheduleResult{
			SuggestedHost:  filteredNodes[0].Name,
			EvaluatedNodes: 1 + len(failedPredicateMap),
			FeasibleNodes:  1,
		}, nil
	}
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

--- a/pkg/scheduler/core/extender.go
+++ b/pkg/scheduler/core/extender.go
@@ -253,8 +253,8 @@ func convertToNodeNameToVictims(
 }
 
 // Filter based on extender implemented predicate functions. The filtered list is
-// expected to be a subset of the supplied list. failedNodesMap optionally contains
-// the list of failed nodes and failure reasons.
+// expected to be a subset of the supplied list; otherwise the function returns an error.
+// failedNodesMap optionally contains the list of failed nodes and failure reasons.
 func (h *HTTPExtender) Filter(
 	pod *v1.Pod,
 	nodes []*v1.Node, nodeNameToInfo map[string]*schedulernodeinfo.NodeInfo,
@@ -298,14 +298,20 @@ func (h *HTTPExtender) Filter(
 	}
 
 	if h.nodeCacheCapable && result.NodeNames != nil {
-		nodeResult = make([]*v1.Node, 0, len(*result.NodeNames))
-		for i := range *result.NodeNames {
-			nodeResult = append(nodeResult, nodeNameToInfo[(*result.NodeNames)[i]].Node())
+		nodeResult = make([]*v1.Node, len(*result.NodeNames))
+		for i, nodeName := range *result.NodeNames {
+			if node, ok := nodeNameToInfo[nodeName]; ok {
+				nodeResult[i] = node.Node()
+			} else {
+				return nil, nil, fmt.Errorf(
+					"extender %q claims a filtered node %q which is not found in nodeNameToInfo map",
+					h.extenderURL, nodeName)
+			}
 		}
 	} else if result.Nodes != nil {
-		nodeResult = make([]*v1.Node, 0, len(result.Nodes.Items))
+		nodeResult = make([]*v1.Node, len(result.Nodes.Items))
 		for i := range result.Nodes.Items {
-			nodeResult = append(nodeResult, &result.Nodes.Items[i])
+			nodeResult[i] = &result.Nodes.Items[i]
 		}
 	}
 

--- a/pkg/scheduler/scheduler.go
+++ b/pkg/scheduler/scheduler.go
@@ -304,7 +304,7 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
 
 	node, victims, nominatedPodsToClear, err := sched.config.Algorithm.Preempt(preemptor, sched.config.NodeLister, scheduleErr)
 	if err != nil {
-		klog.Errorf("Error preempting victims to make room for %v/%v.", preemptor.Namespace, preemptor.Name)
+		klog.Errorf("Error preempting victims to make room for %v/%v: %v", preemptor.Namespace, preemptor.Name, err)
 		return "", err
 	}
 	var nodeName = ""

